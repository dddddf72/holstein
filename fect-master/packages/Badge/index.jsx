import { computed, ref } from 'vue'
import { createNameSpace, theme, validator, useProvider } from '../utils'
const [createComponent] = createNameSpace('Badge')
import './badge.less'

const { normalTypes, normalSizes } = theme

const READONLY_BADGE_ANCHOR_KEY = 'badgeAnchorKey'

const queryBgColor = (type) => {
  const bgsPool = {
    default: 'var(--primary-foreground)',
    success: 'var(--success-default)',
    warning: 'var(--warning-default)',
    error: 'var(--error-default)',
  }
  return bgsPool[type]
}

const queryFontSize = (size) => {
  const sizesPool = {
    mini: '11px',
    small: '12px',
    medium: '14px',
    large: '16px',
  }
  return sizesPool[size]
}

export default createComponent({
  props: {
    type: {
      type: String,
      default: 'default',
      validator: validator.enums(normalTypes),
    },
    size: {
      type: String,
      default: 'medium',
      validator: validator.enums(normalSizes),
    },
    dot: Boolean,
  },
  setup(props, { attrs, slots }) {
    const { ctx } = useProvider(READONLY_BADGE_ANCHOR_KEY)
    const hasParent = ref(Boolean(ctx))
    const calcStyle = computed(() => {
      const styles = {
        backgroundColor: queryBgColor(props.type),
        fontSize: queryFontSize(props.size),
      }
      return styles
    })

    const palcementStyle = computed(() => {
      const { top, left, right, value, origin, bottom } = ctx
      const style = {
        position: 'absolute',
        top: `${top || 'auto'}`,
        left: `${left || 'auto'}`,
        right: `${right || 'auto'}`,
        bottom: `${bottom || 'auto'}`,
        transform: `${value}`,
        transformOrigin: `${origin}`,
        zIndex: 1,
      }
      return style
    })

    const renderElement = () => {
      return (
        <span
          {...attrs}
          class={`${props.dot ? 'fect-dot' : ''} fect-badge`}
          style={calcStyle.value}
        >
          {!props.dot && slots.default?.()}
        </span>
      )
    }
    return () => (
      <>
        {hasParent.value ? (
          <sup style={palcementStyle.value}>{renderElement()}</sup>
        ) : (
          <>{renderElement()}</>
        )}
      </>
    )
  },
})
