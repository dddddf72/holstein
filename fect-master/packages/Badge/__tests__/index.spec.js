import { mount } from '@vue/test-utils'
import Badge from '../index'

describe('Badge', () => {
  it('should be render as element', () => {
    const badge = mount(<Badge />)
    expect(() => badge.unmount()).not.toThrow()
  })
  it('should be support different types', () => {
    const wrapper = mount({
      setup() {
        return { types: ['default', 'success', 'warning', 'error'] }
      },
      render() {
        return (
          <>
            {this.types.map((type) => (
              <Badge key={type} type={type} type-data={type}>
                {type}
              </Badge>
            ))}
          </>
        )
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    const els = wrapper.findAll('.fect-badge')
    expect(els[0].attributes('type-data')).toBe('default')
    expect(els[1].attributes('type-data')).toBe('success')
    expect(els[2].attributes('type-data')).toBe('warning')
    expect(els[3].attributes('type-data')).toBe('error')
  })
  it('should be support different size', () => {
    const wrapper = mount({
      setup() {
        return { sizes: ['mini', 'small', 'medium', 'large'] }
      },
      render() {
        return (
          <>
            {this.sizes.map((type) => (
              <Badge key={type} size={type}>
                {type}
              </Badge>
            ))}
          </>
        )
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
    const els = wrapper.findAll('.fect-badge')
    expect(els[0].attributes('style')).toBe('font-size: 11px;')
    expect(els[1].attributes('style')).toBe('font-size: 12px;')
    expect(els[2].attributes('style')).toBe('font-size: 14px;')
    expect(els[3].attributes('style')).toBe('font-size: 16px;')
  })
  it('should render as a dot element and innerText will be null', () => {
    const dot = mount(<Badge dot>dot1234</Badge>)
    expect(dot.html()).toMatchSnapshot()
    expect(dot.find('.fect-dot')).toBeTruthy()
    expect(dot.find('.fect-dot').text()).toBe('')
  })
})
