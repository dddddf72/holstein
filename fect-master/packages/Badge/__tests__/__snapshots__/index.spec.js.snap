// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should be support different size 1`] = `<span class="fect-badge" style="font-size: 11px;">mini</span><span class="fect-badge" style="font-size: 12px;">small</span><span class="fect-badge" style="font-size: 14px;">medium</span><span class="fect-badge" style="font-size: 16px;">large</span>`;

exports[`Badge should be support different types 1`] = `<span type-data="default" class="fect-badge" style="font-size: 14px;">default</span><span type-data="success" class="fect-badge" style="font-size: 14px;">success</span><span type-data="warning" class="fect-badge" style="font-size: 14px;">warning</span><span type-data="error" class="fect-badge" style="font-size: 14px;">error</span>`;

exports[`Badge should render as a dot element and innerText will be null 1`] = `<span class="fect-dot fect-badge" style="font-size: 14px;"><!----></span>`;
